ngraph/core/reference/include/ngraph/runtime/reference/product.hpp

```diff
@@ -16,29 +17,27 @@ namespace ngraph
         namespace reference
         {
             template <typename T>
-            void product(const T* arg,
-                         T* out,
-                         const Shape& in_shape,
-                         const AxisSet& reduction_axes,
-                         bool keep_dims)
+            void product(const T* arg, T* out, const Shape& in_shape, const AxisSet& reduction_axes)
             {
-                auto out_shape = reduce(in_shape, reduction_axes, keep_dims);
-                CoordinateTransform output_transform(out_shape);
+                constexpr bool dont_keep_dims_in_output = false;
+                auto out_shape = reduce(in_shape, reduction_axes, dont_keep_dims_in_output);
+                std::fill(out, out + shape_size(out_shape), 1);
 
-                for (const Coordinate& output_coord : output_transform)
-                {
-                    out[output_transform.index(output_coord)] = 1;
-                }
-
-                CoordinateTransform input_transform(in_shape);
+                const auto in_strides = row_major_strides(in_shape);
+                const auto out_strides = row_major_strides(out_shape);
 
+                CoordinateTransformBasic input_transform(in_shape);
                 for (const Coordinate& input_coord : input_transform)
                 {
-                    Coordinate output_coord = reduce(input_coord, reduction_axes, keep_dims);
+                    Coordinate output_coord =
+                        reduce(input_coord, reduction_axes, dont_keep_dims_in_output);
 
-                    size_t output_index = output_transform.index(output_coord);
+                    size_t in_idx = std::inner_product(
+                        input_coord.begin(), input_coord.end(), in_strides.begin(), 0);
+                    size_t out_idx = std::inner_product(
+                        output_coord.begin(), output_coord.end(), out_strides.begin(), 0);
```

Comment:
You can consider adding `const` for in_idx and out_idx

Reply:
Why not `output_coord` as well? ;)

Reply:
no objections 😄

Reply:
Done